plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -273,79 +279,185 @@ private static boolean isDirectory(URI uri)
+        private final int bufferSize;
 
         @GuardedBy("this")
-        private InputStreamSliceInput sliceInput;
+        private ExchangeSourceFile currentFile;
+        @GuardedBy("this")
+        private long fileOffset;
         @GuardedBy("this")
-        private boolean closed;
+        private SliceInput sliceInput;
+        @GuardedBy("this")
+        private int sliceSize = -1;
+        private volatile boolean closed;
+        private volatile long bufferRetainedSize;
+        private volatile ListenableFuture<Void> inProgressReadFuture = immediateVoidFuture();
 
         public AzureExchangeStorageReader(
-                BlobServiceClient blobServiceClient,
-                Queue<ExchangeSourceFile> sourceFiles)
+                BlobServiceAsyncClient blobServiceAsyncClient,
+                Queue<ExchangeSourceFile> sourceFiles,
+                int blockSize,
+                int maxPageStorageSize)
         {
-            this.blobServiceClient = requireNonNull(blobServiceClient, "blobServiceClient is null");
+            this.blobServiceAsyncClient = requireNonNull(blobServiceAsyncClient, "blobServiceAsyncClient is null");
             this.sourceFiles = requireNonNull(sourceFiles, "sourceFiles is null");
+            this.blockSize = blockSize;
+            // Make sure buffer can accommodate at least one complete Slice, and keep reads aligned to block boundaries
+            this.bufferSize = maxPageStorageSize + blockSize;
+
+            // Safe publication of S3ExchangeStorageReader is required as it's a mutable class
+            fillBuffer();
         }
 
         @Override
         public synchronized Slice read()
                 throws IOException
         {
-            if (closed) {
+            if (closed || !inProgressReadFuture.isDone()) {
                 return null;
             }
-            if (sliceInput != null && sliceInput.isReadable()) {
-                return sliceInput.readSlice(sliceInput.readInt());
+
+            try {
+                getFutureValue(inProgressReadFuture);
+            }
+            catch (RuntimeException e) {
+                throw new IOException(e);
             }
-            ExchangeSourceFile sourceFile = sourceFiles.poll();
-            if (sourceFile != null) {
-                String containerName = getContainerName(sourceFile.getFileUri());
-                String blobName = getPath(sourceFile.getFileUri());
-                Optional<SecretKey> secretKey = sourceFile.getSecretKey();
-                BlobClient blobClient = blobServiceClient.getBlobContainerClient(containerName).getBlobClient(blobName);
-                if (secretKey.isPresent()) {
-                    blobClient = blobClient.getCustomerProvidedKeyClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+
+            if (sliceSize < 0) {
+                sliceSize = sliceInput.readInt();
+            }
+            Slice data = sliceInput.readSlice(sliceSize);
```

Comment:
It is not obvious that `sliceInput` will have at least `sliceSize` bytes at this point (after single call to `fillBuffer`) worth a comment.

Reply:
I agree, but I find it a little awkward to add comment here as the code seems to explain itself better.